# CoreCrypto Transactions - Performance Improvements

## Current context
After the [initial event performance investigation](../event-performance-investigation/PerformanceImprovements.ipynb), we have cut processing time.
However, the decryption of events still takes ~37% of event processing. It is the biggest single time consumption that is common to all messages.

By decryption, we mean something like `proteusClient.decryptMessage(session,data)`, or `mlsClient.decryptMessage(groupId, message)`, which internally
writes data to the filesystem to store the new cryptographic state of the relevant Proteus session or MLS group.

Writing to the filesystem operations is... _s l o w_.

## Improvements

The CoreCrypto team has been working on improving the performance of the decryption process by introducing a transaction API.
The theory is that by using transactions, we can reduce the number of filesystem operations, decrypting multiple messages and reserving a single file write operation at the end.

Adapting to the new transaction format required intense changing in the way we handle events, and we're here looking for a before/after comparison.

## Data collection

Data was collected by running the [PocIntegrationTest](../../../tango-tests/src/integrationTest/kotlin/PocIntegrationTest.kt) test suite before and after the changes to decrypt multiple messages in a single transaction, and the results are in the neighbour [proteus_benchmark_summary.csv](./proteus_benchmark_summary.csv) file.

#### Conditions of the test
The tests are completely end-to-end, with a mocked network layer.
This means that the test simulates a user logging in, doing initial sync, etc. and we have a whole proper client setup.
We begin measuring the time only after this initial setup is done, and we are ready to start processing events.

Here we begin showing and exploring the collected data:

In [77]:
%use kandy
%use dataframe

@file:DependsOn("org.jetbrains.kotlinx:kotlinx-datetime:0.6.1")

val proteusBenchmarkSummary = DataFrame.readCsv("proteus_benchmark_summary.csv", delimiter = ',')
proteusBenchmarkSummary

<iframe onload="o_resize_iframe_out_13()" style="width:100%;" class="result_container" id="iframe_out_13" frameBorder="0" srcdoc=" <html theme='dark'>
 <head>
 <style type="text/css">
 :root {
 --background: #fff;
 --background-odd: #f5f5f5;
 --background-hover: #d9edfd;
 --header-text-color: #474747;
 --text-color: #848484;
 --text-color-dark: #000;
 --text-color-medium: #737373;
 --text-color-pale: #b3b3b3;
 --inner-border-color: #aaa;
 --bold-border-color: #000;
 --link-color: #296eaa;
 --link-color-pale: #296eaa;
 --link-hover: #1a466c;
}

:root[theme="dark"], :root [data-jp-theme-light="false"], .dataframe_dark{
 --background: #303030;
 --background-odd: #3c3c3c;
 --background-hover: #464646;
 --header-text-color: #dddddd;
 --text-color: #b3b3b3;
 --text-color-dark: #dddddd;
 --text-color-medium: #b2b2b2;
 --text-color-pale: #737373;
 --inner-border-color: #707070;
 --bold-border-color: #777777;
 --link-color: #008dc0;
 --link-color-pale: #97e1fb;
 --link-hover: #00688e;
}

p.dataframe_description {
 color: var(--text-color-dark);
}

table.dataframe {
 font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
 font-size: 12px;
 background-color: var(--background);
 color: var(--text-color-dark);
 border: none;
 border-collapse: collapse;
}

table.dataframe th, td {
 padding: 6px;
 border: 1px solid transparent;
 text-align: left;
}

table.dataframe th {
 background-color: var(--background);
 color: var(--header-text-color);
}

table.dataframe td {
 vertical-align: top;
 white-space: nowrap;
}

table.dataframe th.bottomBorder {
 border-bottom-color: var(--bold-border-color);
}

table.dataframe tbody > tr:nth-child(odd) {
 background: var(--background-odd);
}

table.dataframe tbody > tr:nth-child(even) {
 background: var(--background);
}

table.dataframe tbody > tr:hover {
 background: var(--background-hover);
}

table.dataframe a {
 cursor: pointer;
 color: var(--link-color);
 text-decoration: none;
}

table.dataframe tr:hover > td a {
 color: var(--link-color-pale);
}

table.dataframe a:hover {
 color: var(--link-hover);
 text-decoration: underline;
}

table.dataframe img {
 max-width: fit-content;
}

table.dataframe th.complex {
 background-color: var(--background);
 border: 1px solid var(--background);
}

table.dataframe .leftBorder {
 border-left-color: var(--inner-border-color);
}

table.dataframe .rightBorder {
 border-right-color: var(--inner-border-color);
}

table.dataframe .rightAlign {
 text-align: right;
}

table.dataframe .expanderSvg {
 width: 8px;
 height: 8px;
 margin-right: 3px;
}

table.dataframe .expander {
 display: flex;
 align-items: center;
}

/* formatting */

table.dataframe .null {
 color: var(--text-color-pale);
}

table.dataframe .structural {
 color: var(--text-color-medium);
 font-weight: bold;
}

table.dataframe .dataFrameCaption {
 font-weight: bold;
}

table.dataframe .numbers {
 color: var(--text-color-dark);
}

table.dataframe td:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}

table.dataframe tr:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}


:root {
 --scroll-bg: #f5f5f5;
 --scroll-fg: #b3b3b3;
}
:root[theme="dark"], :root [data-jp-theme-light="false"]{
 --scroll-bg: #3c3c3c;
 --scroll-fg: #97e1fb;
}
body {
 scrollbar-color: var(--scroll-fg) var(--scroll-bg);
}
body::-webkit-scrollbar {
 width: 10px; /* Mostly for vertical scrollbars */
 height: 10px; /* Mostly for horizontal scrollbars */
}
body::-webkit-scrollbar-thumb {
 background-color: var(--scroll-fg);
}
body::-webkit-scrollbar-track {
 background-color: var(--scroll-bg);
}
 </style>
 </head>
 <body>
 <table class="dataframe" id="df_-385875944"></table>

<p class="dataframe_description">DataFrame: rowsCount = 3, columnsCount = 3</p>

 </body>
 <script>
 (function () {
 window.DataFrame = window.DataFrame || new (function () {
 this.addTable = function (df) {
 let cols = df.cols;
 for (let i = 0; i < cols.length; i++) {
 for (let c of cols[i].children) {
 cols[c].parent = i;
 }


In [78]:
import org.jetbrains.letsPlot.core.spec.plotson.scale

fun plotRow(row: DataRow<Line_104_jupyter._DataFrameType2>) = plot {
    bars {
        layout.title = "Time to process ${row.Scenario}, lower is better"
        y(listOf(row.`Before (ms)`, row.`After (ms)`), name = "ms")
        x(listOf("Before", "With Transactions"))
        fillColor("x") {
            scale = continuous(Color.GREEN..Color.LIGHT_GREEN)
        }
    }
}
plotRow(proteusBenchmarkSummary[0])

<head>
 <meta charset="UTF-8">
 <style> html, body { margin: 0; overflow: hidden; } </style>
 <script type="text/javascript" data-lets-plot-script="library" src="https://cdn.jsdelivr.net/gh/JetBrains/lets-plot@v4.5.1/js-package/distr/lets-plot.min.js"></script>
 </head>
 <body>
 <div id="H0EDiA"></div>
 <script type="text/javascript" data-lets-plot-script="plot">
 
 (function() {
 // ----------
 
 var plotSpec={
"ggtitle":{
"text":"Time to process 2000 messages from 1 conversation, lower is better"
},
"mapping":{
},
"data":{
},
"kind":"plot",
"scales":[{
"aesthetic":"y",
"limits":[null,null]
},{
"aesthetic":"x",
"discrete":true
},{
"aesthetic":"fill",
"scale_mapper_kind":"color_gradient",
"high":"#91cc75",
"low":"#3ba272",
"limits":[null,null]
}],
"layers":[{
"mapping":{
"y":"ms",
"x":"x",
"fill":"x"
},
"stat":"identity",
"data":{
"ms":[10218.0,6046.0],
"x":["Before","With Transactions"]
},
"sampling":"none",
"inherit_aes":false,
"position":"dodge",
"geom":"bar",
"data_meta":{
"series_annotations":[{
"type":"int",
"column":"ms"
},{
"type":"str",
"column":"x"
}]
}
}],
"spec_id":"179"
};
 var containerDiv = document.getElementById("H0EDiA");
 
 var toolbar = null;
 var plotContainer = containerDiv; 
 
 var options = {
 sizing: {
 width_mode: "fixed",
 height_mode: "fixed",
 width: 600.0,
 height: 400.0
 }
 };
 var fig = LetsPlot.buildPlotFromProcessedSpecs(plotSpec, -1, -1, plotContainer, options);
 if (toolbar) {
 toolbar.bind(fig);
 }
 
 // ----------
 })();
 
 </script>
 </body>
</html>"> 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 Before 
 
 
 
 
 
 
 
 
 With Transactions 
 
 
 
 
 
 
 
 
 
 
 0 
 
 
 
 
 
 
 2,000 
 
 
 
 
 
 
 4,000 
 
 
 
 
 
 
 6,000 
 
 
 
 
 
 
 8,000 
 
 
 
 
 
 
 10,000 
 
 
 
 
 
 
 
 
 Time to process 2000 messages from 1 conversation, lower is better 
 
 
 
 
 ms 
 
 
 
 
 x 
 
 
 
 
 
 
 
 
 x 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 Before 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 With Transactions

In [79]:
plotRow(proteusBenchmarkSummary[1])

<head>
 <meta charset="UTF-8">
 <style> html, body { margin: 0; overflow: hidden; } </style>
 <script type="text/javascript" data-lets-plot-script="library" src="https://cdn.jsdelivr.net/gh/JetBrains/lets-plot@v4.5.1/js-package/distr/lets-plot.min.js"></script>
 </head>
 <body>
 <div id="9O9VTO"></div>
 <script type="text/javascript" data-lets-plot-script="plot">
 
 (function() {
 // ----------
 
 var plotSpec={
"ggtitle":{
"text":"Time to process 1000 messages across 5 conversations, lower is better"
},
"mapping":{
},
"data":{
},
"kind":"plot",
"scales":[{
"aesthetic":"y",
"limits":[null,null]
},{
"aesthetic":"x",
"discrete":true
},{
"aesthetic":"fill",
"scale_mapper_kind":"color_gradient",
"high":"#91cc75",
"low":"#3ba272",
"limits":[null,null]
}],
"layers":[{
"mapping":{
"y":"ms",
"x":"x",
"fill":"x"
},
"stat":"identity",
"data":{
"ms":[5025.0,2920.0],
"x":["Before","With Transactions"]
},
"sampling":"none",
"inherit_aes":false,
"position":"dodge",
"geom":"bar",
"data_meta":{
"series_annotations":[{
"type":"int",
"column":"ms"
},{
"type":"str",
"column":"x"
}]
}
}],
"spec_id":"182"
};
 var containerDiv = document.getElementById("9O9VTO");
 
 var toolbar = null;
 var plotContainer = containerDiv; 
 
 var options = {
 sizing: {
 width_mode: "fixed",
 height_mode: "fixed",
 width: 600.0,
 height: 400.0
 }
 };
 var fig = LetsPlot.buildPlotFromProcessedSpecs(plotSpec, -1, -1, plotContainer, options);
 if (toolbar) {
 toolbar.bind(fig);
 }
 
 // ----------
 })();
 
 </script>
 </body>
</html>"> 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 Before 
 
 
 
 
 
 
 
 
 With Transactions 
 
 
 
 
 
 
 
 
 
 
 0 
 
 
 
 
 
 
 1,000 
 
 
 
 
 
 
 2,000 
 
 
 
 
 
 
 3,000 
 
 
 
 
 
 
 4,000 
 
 
 
 
 
 
 5,000 
 
 
 
 
 
 
 
 
 Time to process 1000 messages across 5 conversations, lower is better 
 
 
 
 
 ms 
 
 
 
 
 x 
 
 
 
 
 
 
 
 
 x 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 Before 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 With Transactions

In [80]:
plotRow(proteusBenchmarkSummary[2])

<head>
 <meta charset="UTF-8">
 <style> html, body { margin: 0; overflow: hidden; } </style>
 <script type="text/javascript" data-lets-plot-script="library" src="https://cdn.jsdelivr.net/gh/JetBrains/lets-plot@v4.5.1/js-package/distr/lets-plot.min.js"></script>
 </head>
 <body>
 <div id="77ELCY"></div>
 <script type="text/javascript" data-lets-plot-script="plot">
 
 (function() {
 // ----------
 
 var plotSpec={
"ggtitle":{
"text":"Time to process 5000 messages across 6 conversations, lower is better"
},
"mapping":{
},
"data":{
},
"kind":"plot",
"scales":[{
"aesthetic":"y",
"limits":[null,null]
},{
"aesthetic":"x",
"discrete":true
},{
"aesthetic":"fill",
"scale_mapper_kind":"color_gradient",
"high":"#91cc75",
"low":"#3ba272",
"limits":[null,null]
}],
"layers":[{
"mapping":{
"y":"ms",
"x":"x",
"fill":"x"
},
"stat":"identity",
"data":{
"ms":[24768.0,13717.0],
"x":["Before","With Transactions"]
},
"sampling":"none",
"inherit_aes":false,
"position":"dodge",
"geom":"bar",
"data_meta":{
"series_annotations":[{
"type":"int",
"column":"ms"
},{
"type":"str",
"column":"x"
}]
}
}],
"spec_id":"185"
};
 var containerDiv = document.getElementById("77ELCY");
 
 var toolbar = null;
 var plotContainer = containerDiv; 
 
 var options = {
 sizing: {
 width_mode: "fixed",
 height_mode: "fixed",
 width: 600.0,
 height: 400.0
 }
 };
 var fig = LetsPlot.buildPlotFromProcessedSpecs(plotSpec, -1, -1, plotContainer, options);
 if (toolbar) {
 toolbar.bind(fig);
 }
 
 // ----------
 })();
 
 </script>
 </body>
</html>"> 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 Before 
 
 
 
 
 
 
 
 
 With Transactions 
 
 
 
 
 
 
 
 
 
 
 0 
 
 
 
 
 
 
 5,000 
 
 
 
 
 
 
 10,000 
 
 
 
 
 
 
 15,000 
 
 
 
 
 
 
 20,000 
 
 
 
 
 
 
 25,000 
 
 
 
 
 
 
 
 
 Time to process 5000 messages across 6 conversations, lower is better 
 
 
 
 
 ms 
 
 
 
 
 x 
 
 
 
 
 
 
 
 
 x 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 Before 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 With Transactions

## Analysis
It's easy to see that the graphs look very similar.
1. The number of messages seems to affect directly the time it takes to process them, which seems logical.
2. The new approach seems consistently faster

In [81]:
val improvementPercentages = proteusBenchmarkSummary.map { row ->
    1 - (row.`After (ms)`.toDouble() / row.`Before (ms)`.toDouble())
}
plot {
    points {
        x(improvementPercentages, "Improvement %")
        x.axis.limits = 0.0..1.0
        y(proteusBenchmarkSummary.map { row -> row.Scenario }, "Scenario")
        color = Color.LIGHT_GREEN
    }
}

<head>
 <meta charset="UTF-8">
 <style> html, body { margin: 0; overflow: hidden; } </style>
 <script type="text/javascript" data-lets-plot-script="library" src="https://cdn.jsdelivr.net/gh/JetBrains/lets-plot@v4.5.1/js-package/distr/lets-plot.min.js"></script>
 </head>
 <body>
 <div id="xZGJDy"></div>
 <script type="text/javascript" data-lets-plot-script="plot">
 
 (function() {
 // ----------
 
 var plotSpec={
"mapping":{
},
"coord":{
"name":"cartesian",
"xlim":[0.0,1.0],
"flip":false
},
"data":{
},
"kind":"plot",
"scales":[{
"aesthetic":"x",
"limits":[null,null]
},{
"aesthetic":"x",
"limits":[null,null]
},{
"aesthetic":"y",
"discrete":true
}],
"layers":[{
"mapping":{
"x":"Improvement %",
"y":"Scenario"
},
"stat":"identity",
"data":{
"Improvement %":[0.4082990800548052,0.4189054726368159,0.4461805555555556],
"Scenario":["2000 messages from 1 conversation","1000 messages across 5 conversations","5000 messages across 6 conversations"]
},
"color":"#91cc75",
"sampling":"none",
"inherit_aes":false,
"position":"identity",
"geom":"point",
"data_meta":{
"series_annotations":[{
"type":"float",
"column":"Improvement %"
},{
"type":"str",
"column":"Scenario"
}]
}
}],
"spec_id":"188"
};
 var containerDiv = document.getElementById("xZGJDy");
 
 var toolbar = null;
 var plotContainer = containerDiv; 
 
 var options = {
 sizing: {
 width_mode: "fixed",
 height_mode: "fixed",
 width: 600.0,
 height: 400.0
 }
 };
 var fig = LetsPlot.buildPlotFromProcessedSpecs(plotSpec, -1, -1, plotContainer, options);
 if (toolbar) {
 toolbar.bind(fig);
 }
 
 // ----------
 })();
 
 </script>
 </body>
</html>"> 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 0.0 
 
 
 
 
 
 
 
 
 0.5 
 
 
 
 
 
 
 
 
 1.0 
 
 
 
 
 
 
 
 
 
 
 2000 messages from 1 conversation 
 
 
 
 
 
 
 1000 messages across 5 conversations 
 
 
 
 
 
 
 5000 messages across 6 conversations 
 
 
 
 
 
 
 
 
 Scenario 
 
 
 
 
 Improvement %

In [82]:
import java.text.NumberFormat

val formatter = NumberFormat.getPercentInstance()
    formatter.maximumFractionDigits = 2
formatter.format(improvementPercentages.average())

42,45 %

# Results

So yeah, it looks like a quick 42% improvement across the board? Yey :)

Kudos to everyone involved.